rust/src/storage/s3/mod.rs

```diff
@@ -856,6 +851,37 @@ impl dyn LockClient {
     }
 }
 
+const S3_GET_INTERNAL_SERVER_ERROR_RETRIES: usize = 10;
+
+async fn get_object_with_retries(
+    client: &S3Client,
+    bucket: &str,
+    key: &str,
+) -> Result<GetObjectOutput, RusotoError<GetObjectError>> {
+    let mut tries = 0;
+    loop {
+        let result = client
+            .get_object(GetObjectRequest {
+                bucket: bucket.to_string(),
+                key: key.to_string(),
+                ..Default::default()
+            })
+            .await;
+        match result {
+            Err(RusotoError::Unknown(e))
+                if e.status.is_server_error() && tries < S3_GET_INTERNAL_SERVER_ERROR_RETRIES =>
+            {
+                log::warn!("Got {:?}, retrying", e);
+                tries += 1;
+                continue;
```

Comment:
WDYT of having a tiny sleep before retrying right after? Retry with the exponential backoff would be perfect.

Reply:
The error is very rare, like 5-6 times per day for 2-3 topics with 5-10k msg/s rate. And it usually was resolved with single retry, maybe I saw once two. So I figured that the time in between would be redundant. But again, that's just for our use case

Reply:
Oh cool ty, will take a look at it tomorrow